File: apps/api/src/schemas/iab_gvl.py
"""Read-only response schemas for the IAB GVL endpoints."""

from __future__ import annotations

from datetime import datetime
from typing import Any

from pydantic import BaseModel, ConfigDict


class GvlMetaResponse(BaseModel):
    """Currently-cached GVL version metadata."""

    model_config = ConfigDict(from_attributes=True)

    gvl_specification_version: int
    vendor_list_version: int
    tcf_policy_version: int
    last_updated: datetime
    synced_at: datetime


class IabVendorResponse(BaseModel):
    """A single IAB-registered vendor."""

    model_config = ConfigDict(from_attributes=True)

    id: int
    name: str
    purposes: list[int] | None = None
    leg_int_purposes: list[int] | None = None
    flexible_purposes: list[int] | None = None
    special_purposes: list[int] | None = None
    features: list[int] | None = None
    special_features: list[int] | None = None
    policy_url: str | None = None
    deleted_date: datetime | None = None
    uses_cookies: bool | None = None
    cookie_refresh: bool | None = None
    uses_non_cookie_access: bool | None = None
    cookie_max_age_seconds: int | None = None
    data_retention: dict[str, Any] | None = None
    urls: list[dict[str, Any]] | None = None
    data_declaration: list[int] | None = None


class IabVendorListResponse(BaseModel):
    """Paginated list of vendors."""

    items: list[IabVendorResponse]
    total: int
    limit: int
    offset: int

File: apps/admin-ui/src/api/iab-gvl.ts
import apiClient from './client';

/** Singleton metadata for the currently-cached IAB Global Vendor List. */
export interface GvlMeta {
  gvl_specification_version: number;
  vendor_list_version: number;
  tcf_policy_version: number;
  last_updated: string;
  synced_at: string;
}

/** A single IAB-registered vendor as cached locally from the GVL. */
export interface IabVendor {
  id: number;
  name: string;
  purposes: number[] | null;
  leg_int_purposes: number[] | null;
  flexible_purposes: number[] | null;
  special_purposes: number[] | null;
  features: number[] | null;
  special_features: number[] | null;
  policy_url: string | null;
  deleted_date: string | null;
  uses_cookies: boolean | null;
  cookie_refresh: boolean | null;
  uses_non_cookie_access: boolean | null;
  cookie_max_age_seconds: number | null;
}

export interface IabVendorListResponse {
  items: IabVendor[];
  total: number;
  limit: number;
  offset: number;
}

export interface ListVendorsParams {
  q?: string;
  include_deleted?: boolean;
  limit?: number;
  offset?: number;
}

export async function getGvlMeta(): Promise<GvlMeta | null> {
  try {
    const { data } = await apiClient.get<GvlMeta>('/iab/gvl-meta');
    return data;
  } catch (err) {
    // 404 means the GVL hasn't been synced yet — surface as null so
    // the UI can render a "not synced" state instead of an error.
    const status = (err as { response?: { status?: number } }).response?.status;
    if (status === 404) return null;
    throw err;
  }
}

export async function listVendors(params: ListVendorsParams = {}): Promise<IabVendorListResponse> {
  const { data } = await apiClient.get<IabVendorListResponse>('/iab/vendors', { params });
  return data;
}

export async function getVendor(id: number): Promise<IabVendor> {
  const { data } = await apiClient.get<IabVendor>(`/iab/vendors/${id}`);
  return data;
}
